File: 백준/Silver/4963. 섬의 개수/섬의 개수.py

from collections import deque

def bfs(i,j):
    queue.append((i,j))
    visited[i][j]=True
    while queue:
        x,y=queue.popleft()
        for k in range(8):
            nx=x+dx[k]
            ny=y+dy[k]
            if 0<=nx<h and 0<=ny<w and visited[nx][ny]==False and mapp[nx][ny]==1:
                visited[nx][ny]=True
                queue.append((nx,ny))


while True:
    w,h=map(int,input().split())
    if w==0 and h==0:
        break
    mapp=[list(map(int,input().split())) for _ in range(h)]
    queue=deque()
    visited=[[False]*w for _ in range(h)]
    cnt=0
    dx=[1,0,-1,0,-1,-1,1,1]
    dy=[0,1,0,-1,-1,1,-1,1]
    for i in range(h):
        for j in range(w):
            if mapp[i][j]==1 and not visited[i][j]:
                bfs(i,j)
                cnt+=1
    print(cnt)
            

File: 백준/Silver/4963. 섬의 개수/섬의 개수.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static StringTokenizer st;
    static int n,m, ans;
    static int[][] mapp;
    static boolean[][] visited;
    static int[] dx ={-1,-1,0,1,1,1,0,-1};
    static int[] dy ={0,-1,-1,-1,0,1,1,1};

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        // 1 : 땅, 0: 바다
        // 가로,세로 대각선 => 걸어갈 수 있음 -> 섬의 개수는?
        while (true) {
            st = new StringTokenizer(br.readLine());
            m = Integer.parseInt(st.nextToken());
            n = Integer.parseInt(st.nextToken());
            if (n==0 && m==0) break;

            mapp = new int[n][m];
            visited = new boolean[n][m];
            for (int i=0; i<n; i++) {
                st = new StringTokenizer(br.readLine());
                for (int j=0; j<m; j++) {
                    mapp[i][j]=Integer.parseInt(st.nextToken());
                }
            }

            int ans = 0; //섬 개수
            for (int i=0; i<n; i++) {
                for (int j=0; j<m; j++) {
                    if (mapp[i][j]==1 && !visited[i][j]) {
                        bfs(i,j);
                        ans++;
                    }
                }
            }
            System.out.println(ans);

        }

    }
    public static void bfs(int x, int y) {
        visited[x][y]=true; //방문체크
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {x,y});

        while (!q.isEmpty()) {
            int[] temp =q.poll();
            int i = temp[0];
            int j = temp[1];

            for (int k=0; k<8; k++) {
                int nx = i+dx[k];
                int ny = j+dy[k];

                if (nx>=0 && nx<n && ny>=0 && ny<m) {
                    if (!visited[nx][ny] && mapp[nx][ny]==1){
                        q.add(new int[] {nx,ny});
                        visited[nx][ny]=true;
                    }
                }
            }
        }


    }

}
